mindmeld/models/embedder_models.py

```diff
@@ -14,119 +14,738 @@
+
+        if as_numpy:
+            return similarity_scores.numpy()
+
+        return similarity_scores
+
+    def find_similarity(self,
+                        src_texts,
+                        tgt_texts=None,
+                        top_n=20,
+                        scores_normalizer=None,
+                        _sim_func=None,
+                        _return_as_dict=False,
+                        _no_sort=False):
+        """Computes the cosine similarity
+
+        Args:
+            src_texts (Union[str, list]): string or list of strings to obtain matching scores for.
+            tgt_texts (list, optional): list of strings that will be matched to.
+                if None, existing cache is used as target strings
+            top_n (int, optional): maximum number of results to populate. if None, equals length
+                of tgt_texts
+            scores_normalizer (str, optional): normalizer type to normalize scores. Allowed values
+                are: "min_max_scaler", "standard_scaler"
+            _sim_func (function, optional): if None, defaults to `_pytorch_cos_sim`. If
+                specified, must take two numpy-array/pytorch-tensor arguments for similarity
+                computation with an optional argument to return results as numpy or tensor
+            _return_as_dict (bool, optional): if the results should be returned as a dictionary of
+                target_text name as keys and scores as corresponding values
+            _no_sort (bool, optional): If True, results are returned without sorting. This is
+                helpful at times when you wish to do additional wrapper operations on top of raw
+                results and would like to save computational time without sorting.
+        Returns:
+            Union[dict, list[tuple]]: if _return_as_dict, returns a dictionary of tgt_texts and
+                their scores, else a list of sorted synonym names paired with their
+                similarity scores (descending order)
+        """
+
+        is_single = False
+        if isinstance(src_texts, str):
+            is_single = True
+            src_texts = [src_texts]
+
+        tgt_texts = [*self.cache.data.keys()] if not tgt_texts else tgt_texts
+        top_n = len(tgt_texts) if not top_n else top_n
+        _sim_func = _sim_func or self._pytorch_cos_sim
+
+        src_vecs = np.asarray(self.get_encodings(list(src_texts), add_to_cache=False))
+        tgt_vecs = np.asarray(self.get_encodings(list(tgt_texts), add_to_cache=False))
+
+        similarity_scores_2d = _sim_func(src_vecs, tgt_vecs, as_numpy=True)  # a 2d numpy array
+
+        results = []
+        for similarity_scores in similarity_scores_2d:
+            similarity_scores = similarity_scores.reshape(-1)
+            # Rounding sometimes helps to bring correct answers on to the list of top scored results
+            similarity_scores = np.around(similarity_scores, decimals=2)
+
+            if scores_normalizer:
```

Comment:
For my understanding, is the normalization of similarity scores here just to make the scores sum to 1 i.e, make it more interpretable?

Reply:
Thats correct! However, we do not use this feature directly in a QA use-case. Because, this feature normalizes across all the synonyms' (whitelist+cnames') scores, however, we need normalization across all doc ids. I left this feature be present so that it can be useful for any future work.